serde-name/src/de_adapter.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Facebook, Inc. and its affiliates
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use serde::de::Visitor;
+
+/// An adapter for `serde::de::Deserializer` implementations that lets you override the
+/// name of the top-level container.
+pub struct DeserializeNameAdapter<D> {
+    inner: D,
+    name: &'static str,
+}
```

Comment:
I guess the most surprising thing to me is that you're having to wrap the Deserializer/Serializer instead of the type you want to rename and this is due to the API of the Deserialize/Serialize traits not giving you this ability.

Reply:
Yes, credits go to @dtolnay for suggesting this approach.